src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs

```diff
@@ -279,38 +282,51 @@ private static void RecordRequestStartEventLog(HttpContext httpContext)
             {
                 return null;
             }
-
             var headers = httpContext.Request.Headers;
-            var requestId = headers.TraceParent;
-            if (requestId.Count == 0)
-            {
-                requestId = headers.RequestId;
-            }
+            _propagator.ExtractTraceIdAndState(headers,
+                static (object? carrier, string fieldName, out string? fieldValue, out IEnumerable<string>? fieldValues) =>
+                {
+                    fieldValue = default;
+                    fieldValues = default;
+                    if (carrier is null)
+                    {
+                        return;
+                    }
+                    var headers = (IHeaderDictionary)carrier;
+                    fieldValue = headers[fieldName];
+                },
+                out var requestId,
+                out var traceState);
 
-            if (!StringValues.IsNullOrEmpty(requestId))
+            if (requestId is not null)
             {
                 activity.SetParentId(requestId);
-                var traceState = headers.TraceState;
-                if (traceState.Count > 0)
-                {
-                    activity.TraceStateString = traceState;
-                }
-
-                // We expect baggage to be empty by default
-                // Only very advanced users will be using it in near future, we encourage them to keep baggage small (few items)
-                var baggage = headers.GetCommaSeparatedValues(HeaderNames.Baggage);
-                if (baggage.Length == 0)
+            }
+            if (traceState is not null)
```

Comment:
>if (traceState is not null) [](http://example.com/codeflow?start=12&length=27)

do we set the trace state if we didn't get request Id? I mean this block should be moved inside the previous block?

Reply:
What does the current propagator implementation do? Do you call the extract delegate for tracestate if you don't find requestid/traceparent?

Reply:
The propagator must try all fields regardless.

https://github.com/dotnet/runtime/blob/b412aac8757d7ac29ab66910e462649eeaecda18/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/LegacyPropagator.cs#L54

My comment here regarding asp.net compatibility with previous behavior. So, regardless of the propagator behavior we should keep asp.net old behavior.

Reply:
Done. Moved the check into the `!string.IsNullOrEmpty(requestId)` block

Reply:
@tarekgh Can you confirm I addressed your concern?

Reply:
Yes, it looks good to me.